service/src/main/java/bio/terra/workspace/app/controller/ReferencedGcpResourceController.java

```diff
@@ -168,8 +174,11 @@ public ResponseEntity<Void> updateBucketObjectReferenceResource(
       if (!StringUtils.isEmpty(objectName)) {
         updateBucketObjectResourceBuilder.fileName(objectName);
       }
-      if (null != cloningInstructions) {
-        updateBucketObjectResourceBuilder.cloningInstructions(cloningInstructions);
+      if (cloningInstructions != null) {
```

Comment:
I know I started this, but...
I would prefer to avoid one-off `toBuilder` uses. I wonder if we could do something like:

Have `ReferencedGcsObjectResource.Builder` have a member that is the `WsmResourceFields.Builder` instead of the `WsmResourceFields` itself. That way both builders would be "open" to take changes. Then the `build()` method of `ReferencedGcsObjectResource.Builder` would call the `build()` method of the `WsmResourceFields.Builder` and generate both objects.

Alternately, maybe the specific resource builder could derive from `WsmResourceFields.Builder` allowing the setters to be flat references, instead of having two separate builders. Since we do not serialize builders, derivation is safe.

wdyt? Maybe that is a separate refactoring ticket?

Reply:
+1. Both suggestions seem doable in this PR (as opposed to ticket)

Reply:
i actually don't quite follow, perhaps we can discuss offline?

I thought toBuilder is generally essential to a Builder class. How do I convert the resource to a builder without it?
re 1: how does this get rid of toBuilder()? This piece would look like this if i change the filed to WsmResourceFields.Builder

```
ReferencedGcsObjectResource.Builder updateBucketObjectResourceBuilder =
          referencedResource.toBuilder();
      if (!StringUtils.isEmpty(bucketName)) {
        updateBucketObjectResourceBuilder.bucketName(bucketName);
      }
      if (!StringUtils.isEmpty(objectName)) {
        updateBucketObjectResourceBuilder.objectName(objectName);
      }
      if (cloningInstructions != null) {
        updateBucketObjectResourceBuilder.resourceCommonFieldsBuilder(
            referencedResource.getWsmResourceFields()
                .cloningInstructions(cloningInstructions));
      }
```

`ReferencedGcsObjectResource#toBuilder()` would look like this:
```
  public Builder toBuilder() {
    return builder()
        .bucketName(getBucketName())
        .objectName(getObjectName())
        .resourceCommonFieldsBuilder(getWsmResourceFieldsBuilder());
  }
```

re 2: What do you mean by derive? `wsmResourceFields.Builder` does not build a resource so specific resource builder can't be a subclass of it.

Reply:
discussed offline with Dan. We will stick with the current way as ResourceBuilder.build() return a resource while ResourceFieldsBuilder.build() returns a resourceFields so they can't have an inheritance relationship.

Reply:
Dan's 1st suggestion is [how proto works](https://developers.google.com/protocol-buffers/docs/reference/java-generated#sub-builders). If Foo has Bar, FooBuilder has BarBuilder.

Currently, ReferencedGcsObjectResource.Builder has WsmResourceField, which looks weird to me. It should have WsmResourceField.Builder.

Could this be

```
        updateBucketObjectResourceBuilder.resourceCommonFieldsBuilder().cloningInstructions(cloningInstructions);

  public Builder toBuilder() {
    return builder()
        .bucketName(getBucketName())
        .objectName(getObjectName())
        .resourceCommonFieldsBuilder(getResourceCommonFieldsBuilder());
  }
```

Happy to chat offline, just wanted to type the code here. :)

Reply:
done.

Reply:
actually i cant.. jackson don't like it if i have a builder as a field. 

https://scans.gradle.com/s/yvs7noqtcfsaq/tests/:service:unitTest/bio.terra.workspace.service.resource.referenced.ReferencedResourceServiceTest$BigQueryReference/bigQueryDataTableReference()?top-execution=1



Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class bio.terra.workspace.service.resource.model.WsmResourceFields$Builder and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: bio.terra.workspace.service.resource.referenced.cloud.gcp.bqdatatable.ReferencedBigQueryDataTableResource["wsmResourceFieldsBuilder"])